src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleControl.cs

```diff
@@ -2575,19 +2569,37 @@ internal static void WriteConsole(ConsoleHandle consoleHandle, ReadOnlySpan<char
             }
 
             outBuffer = output.Slice(cursor);
+            if (!newLine)
+            {
+                WriteConsole(consoleHandle, outBuffer);
+                return;
+            }
+
+            char[] rentedArray = null;
+            string lineEnding = Environment.NewLine;
+            int size = outBuffer.Length + lineEnding.Length;
+
+            // We expect the 'size' will often be small, and thus optimize that case with 'stackalloc'.
+            Span<char> buffer = size <= MaxStackAllocSize ? stackalloc char[size] : default;
 
-            if (newLine)
+            try
             {
-                var endOfLine = Environment.NewLine.AsSpan();
-                var endOfLineLength = endOfLine.Length;
-                Span<char> outBufferLine = stackalloc char[outBuffer.Length + endOfLineLength];
-                outBuffer.CopyTo(outBufferLine);
-                endOfLine.CopyTo(outBufferLine.Slice(outBufferLine.Length - endOfLineLength));
-                WriteConsole(consoleHandle, outBufferLine);
+                if (buffer.IsEmpty)
+                {
+                    rentedArray = ArrayPool<char>.Shared.Rent(size);
+                    buffer = rentedArray.AsSpan().Slice(0, size);
+                }
+
+                outBuffer.CopyTo(buffer);
+                lineEnding.CopyTo(buffer.Slice(outBuffer.Length));
+                WriteConsole(consoleHandle, buffer);
             }
-            else
+            finally
             {
-                WriteConsole(consoleHandle, outBuffer);
+                if (rentedArray is not null)
+                {
+                    ArrayPool<char>.Shared.Return(rentedArray);
```

Comment:
Does use of the shared pool have security implications? Should we be clearing the contents of the buffer before return?

Reply:
By default it is not cleaned
`public override void Return(T[] array, bool clearArray = false)`

Reply:
I don't think there is any security implications of using `ArrayPool.Return` without clearing the array in this case. It's actually pretty common to use it this way in dotnet repo: https://grep.app/search?q=.Shared.Return

Reply:
I agree. Console output doesn't have any sensitive data we should wipe.

Reply:
> I don't think there is any security implications of using `ArrayPool.Return` without clearing the array in this case. It's actually pretty common to use it this way in dotnet repo: https://grep.app/search?q=.Shared.Return

Judging from the grep.app search results it's actually a common pattern to clear the written part of a rented buffer before returning.

Reply:
> I agree. Console output doesn't have any sensitive data we should wipe.

This is arguable, external programs could potentially output sensitive data.

Reply:
> Judging from the grep.app search results it's actually a common pattern to clear the **written part of a rented buffer** before returning.

https://grep.app/search?q=ArrayPool%3Cchar%3E.Shared.Return&filter[repo][0]=dotnet/runtime shows the specific use of `ArrayPool<char>.Shared.Return` in .NET code. I don't see why it matters whether or not the rented array is for writing content out. Take the use of it in [ValueStringBuilder](https://github.com/dotnet/runtime/blob/90ddd5e0f4617fe5b58b59fdf20965483f85fb1f/src/libraries/Common/src/System/Text/ValueStringBuilder.cs#L307-L315) as an example, the content in the rented array could be _sensitive_ I guess, but the code doesn't clear the array when returning.

Also, I did a quick search with the term ["_c# ArrayPool security_"](https://www.google.com/search?q=c%23+ArrayPool+security&source=hp&ei=XKMkY_DXDZHH9APPx5qQBw&iflsig=AJiK0e8AAAAAYySxbLwJV9Ar8g0y47XbXSEqH5D8b4WZ&ved=0ahUKEwjw9bft3Jn6AhWRI30KHc-jBnIQ4dUDCAk&uact=5&oq=c%23+ArrayPool+security&gs_lcp=Cgdnd3Mtd2l6EAMyBQghEKABMgUIIRCgATIFCCEQoAFQAFgAYGxoAHAAeACAAUqIAUqSAQExmAEAoAECoAEB&sclient=gws-wiz), and I don't see anything alarming stands out. So, I think I won't spend more time on this unless it's proven to be unsecure.

Reply:
> > I agree. Console output doesn't have any sensitive data we should wipe.
> 
> This is arguable, external programs could potentially output sensitive data.

It could. But we say about output to _console_ :-) - this says explicitly that no sensitive data in the output should be.

> > I don't think there is any security implications of using `ArrayPool.Return` without clearing the array in this case. It's actually pretty common to use it this way in dotnet repo: https://grep.app/search?q=.Shared.Return
> 
> Judging from the grep.app search results it's actually a common pattern to clear the written part of a rented buffer before returning.

There should be true in argument
https://grep.app/search?q=%5C.Shared%5C.Return.%2Btrue&regexp=true
and the search shows mostly cryptography and serialization code.

Thanks for pointing on possible issue but it is not the case.